File: KSC2015_ProblemSet_n_Solution/Competition2015/01_KRISS/Code/Fortran/mpi/b.f90
!234567890
       program problem_fd
       implicit none
       include 'mpif.h'
       integer i,n1,n2,j,jsta,jend
       integer iter,niter
       integer istatus(MPI_STATUS_SIZE)
       integer ierr,myid,nprocs
       integer iprev,inext,ista,iend
       integer isd1,isd2,irv1,irv2,itag,iroot
       real*8 xi,xf,dx
       real*8 tmr
       real*8, allocatable:: ar(:),br(:)
       real*8, external :: genvv
       real*8 ptmr
       real*8 tic,toc

!  do not change -----{
       n1=1
       n2=100
       n2=100000000
       niter=3
!  do not change -----}

       allocate ( ar(n2), br(n2) )

       xi=0.d0
       xf=1.d0
       dx=(xf-xi)/dble(n2-n1)
       do i=n1,n2
       br(i)=xi+dble(i-n1)*dx
       enddo
       CALL MPI_INIT(ierr)
       tic=MPI_WTIME()
       CALL MPI_COMM_SIZE(MPI_COMM_WORLD,nprocs,ierr)
       CALL MPI_COMM_RANK(MPI_COMM_WORLD,myid,ierr)
       CALL para_range(n1,n2,nprocs,myid,ista,iend)

       write(*,*) 'rank :', myid, 'ista =', ista, 'iend =', iend

       jsta=ista
       jend=iend
       if(myid == 0) jsta=n1+1
       if(myid == nprocs-1) jend=n2-1
       inext=myid+1
       iprev=myid-1
       if(myid == nprocs-1) inext=MPI_PROC_NULL
       if(myid == 0) iprev=MPI_PROC_NULL
       do i=ista,iend
       br(i)=xi+dble(i-n1)*dx
       enddo
       do iter=1,niter
       itag=101
       CALL MPI_ISEND(br(iend), 1,MPI_REAL8, inext,itag,MPI_COMM_WORLD,isd1,ierr)
       CALL MPI_ISEND(br(ista), 1,MPI_REAL8, iprev,itag,MPI_COMM_WORLD,isd2,ierr)
       CALL MPI_IRECV(br(ista-1), 1,MPI_REAL8, iprev,itag,MPI_COMM_WORLD,irv1,ierr)
       CALL MPI_IRECV(br(iend+1), 1,MPI_REAL8, inext,itag,MPI_COMM_WORLD,irv2,ierr)
       CALL MPI_WAIT(isd1,istatus,ierr)
       CALL MPI_WAIT(isd2,istatus,ierr)
       CALL MPI_WAIT(irv1,istatus,ierr)
       CALL MPI_WAIT(irv2,istatus,ierr)
       do j=jsta,jend
!  do not change -----{
       ar(j)=(br(j-1)+br(j+1))/4.d0+br(j)/2.d0+1.d0/genvv(br(j))
!  do not change -----}
       enddo
       do i=ista,iend
!  do not change -----{
       br(i)=ar(i)
!  do not change -----}
       enddo
       enddo
      
       ptmr=0.d0
       do j=jsta,jend
       ptmr=ptmr+ar(j)
       enddo
       iroot=0
       CALL MPI_REDUCE(ptmr, tmr, 1,MPI_REAL8, MPI_SUM,iroot,MPI_COMM_WORLD,ierr)
       if(myid == 0) print *,'tmr = ',tmr
       toc=MPI_WTIME()
       if(myid == 0) print*, toc-tic,' sec'   
   
       deallocate( ar, br )

       CALL MPI_FINALIZE(ierr)
       stop
       end program problem_fd
!234567890
       real*8 function genvv(x)
       implicit none
       real*8 x
       
       genvv=x**2+x**4+x**6+exp(-x**2)+cos(x)+sin(x)+tan(x)
       return
       end
!234567890
       subroutine para_range(n1,n2,nprocs,myid,ista,iend)
       implicit none
       integer n1,n2,nprocs,myid,ista,iend
       integer iwork1,iwork2

       iwork1 = (n2 - n1 + 1) / nprocs
       iwork2 = mod(n2 - n1 + 1, nprocs)
       ista = myid * iwork1 + n1 + min(myid, iwork2)
       iend = ista + iwork1 - 1
       if(iwork2 > myid) iend = iend + 1
       return
       end

File: KSC2015_ProblemSet_n_Solution/Competition2015/01_KRISS/Code/C/mpi/b.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
#include "mpi.h"

#define MAX(a,b) ((a)>(b)?(a):(b))
#define MIN(a,b) ((a)>(b)?(b):(a))

#define TRUE 1
#define FALSE 0


void para_range(int n1,int n2,int nid,int myid,int *ista,int *iend){
	int iwork1, iwork2;
	iwork1 = (n2-n1)/nid;
	iwork2 = (n2-n1)%nid;
	*ista = myid*iwork1 + n1 + MIN(myid,iwork2);
	*iend = *ista + iwork1;
	if(iwork2 > myid) (*iend) ++;
}

double genvv(double x)
{
	double res = x*x + pow(x,4)+ pow(x,6)+exp(-x*x)+cos(x)+sin(x)+tan(x);
	return res;

}

int main (int argc, char **argv)
{
	int i,n1,n2,j,jsta,jend;
	int iter,niter;
	MPI_Status istatus;
	int ierr, myid,nid;
	int iprev, inext, ista, iend;
	MPI_Request isd1,isd2,irv1,irv2;
	int itag, iroot;
	double xi,xf,dx;
	double tmr;
	double *ar, *br;
	double ptmr, tic,toc;

	/* do not change ------ */
	n1 = 0;
	n2 = 100000000;
	niter = 3;
	/* do not change ------ */

	ar = (double*) malloc(n2*sizeof(double));
	br = (double*) malloc(n2*sizeof(double));
	
	xi = 0.L;
	xf = 1.;
	dx = (xf-xi)/(double)(n2-n1-1);

	for(i=n1;i<n2;i++){
		br[i] = xi+(double)(i-n1)*dx;
	}

	MPI_Init(&argc, &argv);
	tic = MPI_Wtime();
	MPI_Comm_size(MPI_COMM_WORLD, &nid);
	MPI_Comm_rank(MPI_COMM_WORLD, &myid);

	para_range(n1,n2,nid,myid,&ista,&iend); // para range를 통해 작업 범위 나눔.
	printf("rank:%10d ista=%15d iend=%15d\n", myid, ista, iend);

	jsta = ista;
	jend = iend;
	if(myid==0) jsta = n1+1;
	if(myid == nid-1) jend = n2-1;

	// send/recv할때 보낼 rank에 사용
	inext = myid + 1; 
	iprev = myid - 1;
	if(myid == nid-1) inext = MPI_PROC_NULL;
	if(myid == 0) iprev = MPI_PROC_NULL;
	for(i=ista;i<iend;i++){
		br[i] = xi+ (double)(i-n1)*dx;
	}

	for(iter=0;iter<niter;iter++){
		itag = 101;
		/**
		 * 각 부분에서 idx-1과 idx+1부분이 필요하기 때문에 
		 * Isend/Irecv로 비동기적으로 보낸다.
		 * Wait를 통해 통신이 동작하는지 확인한다.
		 */
		MPI_Isend(br+iend-1,   1, MPI_DOUBLE, inext, itag, MPI_COMM_WORLD, &isd1); // inext에 b[j-1]전달
		MPI_Isend(br+ista,   1, MPI_DOUBLE, iprev, itag, MPI_COMM_WORLD, &isd2); // iprev b[j+1] 전달
		MPI_Irecv(br+ista-1,   1, MPI_DOUBLE, iprev, itag, MPI_COMM_WORLD, &irv1); // b[j-1] 받음
		MPI_Irecv(br+iend,   1, MPI_DOUBLE, inext, itag, MPI_COMM_WORLD, &irv2); // b[j+1] 받음
		MPI_Wait(&isd1,&istatus);
		MPI_Wait(&isd2,&istatus);
		MPI_Wait(&irv1,&istatus);
		MPI_Wait(&irv2,&istatus);

		for(j=jsta;j<jend;j++) {
			/*  not change -----{ */
			ar[j] = (br[j-1]+br[j+1])/4.L + br[j]/2.L + 1.L/genvv(br[j]);
			/*  not change -----} */
		}
		for(i=ista;i<iend;i++) {
			/*  not change -----{ */
			br[i] = ar[i];
			/*  not change -----} */
		}
	}
	ptmr = 0.L;
	for(j=jsta;j<jend;j++){
		ptmr += ar[j];
	}
	iroot = 0;
	MPI_Reduce(&ptmr, &tmr, 1, MPI_DOUBLE, MPI_SUM, iroot, MPI_COMM_WORLD); // MPI_Reduce로 ptmr 합침
	if(myid==0) printf("tmr = %16.6f\n",tmr);
	toc = MPI_Wtime();
	if(myid==0) printf("%g sec\n",toc-tic);

	free(ar);
	free(br); 

	MPI_Finalize();

}
